File: LeetCode/310. Minimum Height Trees/310.py
from collections import defaultdict, deque

class Solution(object):
    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        if n == 1:
            return [0]
        
        # Graph construction
        graph = defaultdict(set)
        for u, v in edges:
            graph[u].add(v)
            graph[v].add(u)
        
        # Initialize leaves
        leaves = deque([node for node in range(n) if len(graph[node]) == 1])
        
        # Main loop until we have 1 or 2 nodes left
        while n > 2:
            # Exclude current leaves
            n -= len(leaves)
            new_leaves = deque()
            while leaves:
                leaf = leaves.popleft()
                neighbor = graph[leaf].pop()
                graph[neighbor].remove(leaf)
                if len(graph[neighbor]) == 1:
                    new_leaves.append(neighbor)
            leaves = new_leaves
        
        return list(leaves)


def test():
    n = 6
    edges = [[3,0],[3,1],[3,2],[3,4],[5,4]]

    s = Solution()
    print(s.findMinHeightTrees(n, edges))

if __name__ == '__main__':
    test()

File: LeetCode/310. Minimum Height Trees/310.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> result = new ArrayList<>();
        if (n == 1) {
            result.add(0);
            return result;
        }
        
        // Graph construction
        List<Set<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new HashSet<>());
        }
        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }
        
        // Initialize leaves
        List<Integer> leaves = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (graph.get(i).size() == 1) {
                leaves.add(i);
            }
        }
        
        // Main loop until we have 1 or 2 nodes left
        while (n > 2) {
            List<Integer> newLeaves = new ArrayList<>();
            for (int leaf : leaves) {
                int neighbor = graph.get(leaf).iterator().next();
                graph.get(neighbor).remove(leaf);
                if (graph.get(neighbor).size() == 1) {
                    newLeaves.add(neighbor);
                }
            }
            n -= leaves.size();
            leaves = newLeaves;
        }
        
        return leaves;
    }
    public static void main(String[] args) {
        Solution s = new Solution();
        int n = 4;
        int[][] edges = {{1, 0}, {1, 2}, {1, 3}};
        System.out.println(s.findMinHeightTrees(n, edges));
    }
}

